lsp-icons.el

```diff
@@ -0,0 +1,93 @@
+;; MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+;; GNU General Public License for more details.
+
+;; You should have received a copy of the GNU General Public License
+;; along with this program.  If not, see <https://www.gnu.org/licenses/>.
+;;
+;;; Commentary:
+;;
+;;  LSP icons management
+;;
+;;; Code:
+
+(defgroup lsp-icons nil
+  "LSP icons"
+  :group 'lsp-mode
+  :tag "Icons")
+
+(defcustom lsp-headerline-breadcrumb-icons-enable t
+  "If non-nil, icons support is enabled for headerline-breadcrumb."
+  :type 'boolean
+  :group 'lsp-icons)
+
+(declare-function all-the-icons-material "ext:all-the-icons" t t)
+(declare-function lsp-treemacs-symbol-icon "ext:lsp-treemacs" (kind))
+(declare-function lsp-treemacs-get-icon "ext:lsp-treemacs" (icon-name))
+
+(defun lsp-icons--enabled-for-feature (feature)
+  "Check if icons support is enabled for FEATURE."
+  (cond
+   ((= feature 'headerline-breadcrumb) lsp-headerline-breadcrumb-icons-enable)
+   (t t)))
+
+(defun lsp-icons--fix-image-background (image)
+  "Fix IMAGE background if it is a file otherwise return as an icon."
+  (if image
+      (let ((display-image (get-text-property 0 'display image)))
+        (if (and (listp display-image)
+                 (plist-member (cl-copy-list (cl-rest display-image)) :type))
+            (propertize " " 'display
+                        (cl-list* 'image
+                                  (plist-put
+                                   (cl-copy-list
+                                    (cl-rest display-image))
+                                   :background (face-attribute 'header-line :background nil t))))
+          (if (stringp display-image)
+              (replace-regexp-in-string "\s\\|\t" "" display-image)
+            (replace-regexp-in-string "\s\\|\t" "" image))))
+    ""))
+
+(defun lsp-icons-get-by-file-ext (file-ext &optional feature)
+  "Get an icon by file FILE-EXT.
+FEATURE is the feature that will use the icon which we should check
+if its enabled."
+  (when (and file-ext
+             (lsp-icons--enabled-for-feature feature)
+             (require 'lsp-treemacs nil t))
+    (lsp-icons--fix-image-background
+     (lsp-treemacs-get-icon file-ext))))
+
+(defun lsp-icons-get-by-symbol-kind (kind &optional feature)
+  "Get an icon by symbol KIND.
+FEATURE is the feature that will use the icon which we should check
+if its enabled."
+  (when (and kind
+             (lsp-icons--enabled-for-feature feature)
+             (require 'lsp-treemacs nil t))
+    (lsp-icons--fix-image-background
+     (lsp-treemacs-symbol-icon kind))))
+
+(defun lsp-icons-all-the-icons-material-icon (icon-name face fallback &optional feature)
+  "Get a material icon from all-the-icons by ICON-NAME using FACE.
+Fallback to FALLBACK string if not found or not available.
+FEATURE is the feature that will use the icon which we should check
+if its enabled."
+  (if (and (require 'all-the-icons nil t)
```

Comment:
nit: Can we write this slightly differently by put all `require` expression in global scope and check for function available via `fboundp` instead?
That would ensure if the optional library ever change the function name, our package will not be broken

Reply:
Not sure I understand, how would it be?

Reply:
I would suggest using `featurep` instead of `fboundp`? `featurep` is like `require` but it only check for module existence and not going to load it. Anyway, I think @kiennq has the point here so we can keep the code consistent.

```el
(require 'all-the-icons nil t)  ; load `all-the-icons` in global level

(if (featurep 'all-the-icons)
  ;; other execution...
)
```

Reply:
Or better yet, just feature detect the function via `functionp` instead of the package, unless you are using a lot of it.

Reply:
@jcs090218 `featurep` does not check if the package is installed, but rather if it is *loaded*. DOOM lazy-loads `lsp-mode`, for example, so when I start it, `(featurep 'lsp-mode)` yields nil, even though it is installed. `fboundp` is better to check for the existence of packages, because it returns true for `auto-loaded` functions, which works even if the package isn't loaded yet.